MINOR: unify calls to get committed offsets and metadata

mjsax opened the pull request:
This is a follow up to #7304 and #6694 to unify the calls to `consumer.committed()` further and only call it once during startup instead of twice.

mjsax committed:
MINOR: unify calls to get committed offsets and metadata

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -251,24 +246,4 @@ public boolean hasStateStores() {
         return stateMgr.changelogPartitions();
     }
 
-    Map<TopicPartition, Long> committedOffsetForPartitions(final Set<TopicPartition> partitions) {
```
This is moved into `StandbyTask` as we don't share it any longer for both types of tasks.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java:
```diff
@@ -251,7 +251,7 @@ void closeTask(final T task, final boolean clean) {
     }
 
-    private boolean closeUnclean(final T task) {
+    private void closeUnclean(final T task) {
```
side cleanup: return type is never used.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java:
```diff
@@ -110,7 +111,6 @@ boolean hasRunningTasks() {
     void transitionToRunning(final T task) {
         log.debug("Transitioning {} {} to running", taskTypeName, task.id());
         running.put(task.id(), task);
-        task.initializeTaskTime();
```
we remove the method entirely and init task time in the constructor of `StreamTask` (note, that we don't need to init task time for `StandbyTasks` anyway)

mjsax replied:
This in now done implicitly with `resume()`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -251,26 +252,69 @@ public StreamTask(final TaskId id,
+        }
+    }
 
+    private void initializeCommittedOffsets(final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata) {
```
This is a mix of "old" `initializeStateStores` (the part to filter for source topic partitions that are use as changelogs) and "old" `AbstractTaks#committedOffsetForPartitions`)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -251,26 +252,69 @@ public StreamTask(final TaskId id,
+        stateMgr.putOffsetLimits(committedOffsetsForChangelogs);
+    }
+
+    private void initializeTaskTime(final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata) {
```
This is not private. Same implementation as before, however, we get `offsetsAndMetadata` passed in, and don't use the consumer to get them from the brokers.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -251,26 +252,69 @@ public StreamTask(final TaskId id,
+
+
+    @Override
+    public boolean initializeStateStores() {
```
Just simplified -- moved the offset initialization into the constructor.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1243,9 +1243,7 @@ public String toString(final String indent) {
-        final LinkedHashMap<MetricName, Metric> result = new LinkedHashMap<>();
-        result.putAll(adminClientMetrics);
-        return result;
+        return new LinkedHashMap<>(adminClientMetrics);
```
Side cleanup

cpettitt-confluent replied:
Why do we need a LinkedHashMap here? Particularly because the interface doesn't suggest it needs anything more than a Map implementation? Actually I'm also wondering why we need to copy at all since we get back a read only view of the map from Admin and it appears the only consumer of adminClientMetrics immediately copies the results anyway.

mjsax replied:
Not sure about the `LinkedHashMap` -- maybe @guozhangwang knows more?

For why to copy: it's not a public contract that we get a read-only map and hence we need to guard the map from writes via a copy. Not sure if we could rely on an AdminClient implementation detail... \cc @cmccabe ?

cpettitt-confluent replied:
Yes, the only use case I found above this code is to immediately copy it into yet another map (which is why the LinkedHashMap also doesn't make sense), so it should not matter, unless we're expecting to do something else with it down the road. Anyway, per our discussion I'm OK with this going in without fixing it, but it is low hanging fruit to reduce unnecessary work.

mjsax commented:
Call for review @guozhangwang @cadonna @cpettitt-confluent

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -85,14 +90,7 @@ public boolean initializeStateStores() {
-    public void initializeTaskTime() {
-        //no-op
-    }
+    public void initializeTopology() {}
```
Could we also move this only to the `StreamTask`? Doesn't have to be in this PR.

mjsax replied:
Not really, because we call `initializeTopology()` within `AssignedTasks#transitToRunning(T task)` that does not know the type (it generic `T extends Tasks`) -- hence, we would need to do an `instanceof` call within that method what is undesired.

cadonna replied:
Oh, I see. I did not think about that dependency. Nevertheless I think somehow it would be possible to move `initializeTopology()` to `StreamsTask` without an `instanceof` but that needs a bit more thoughts and refactoring and it is out-of-scope for this PR.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -251,26 +252,69 @@ public StreamTask(final TaskId id,
-        log.debug("Initializing state stores");
+        try {
+            final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata = consumer.committed(partitions);
+            initializeCommittedOffsets(offsetsAndMetadata);
+            initializeTaskTime(offsetsAndMetadata);
```
Apparently it is not completely save to call instance methods from the constructor since the object and its member variables are  not completely initialised yet. Static methods would be fine, though.

I do not think that `initializeCommittedOffsets()` is currently problematic since it only accesses member variables from `AbstractTask` which should be already initialised. However, `initializeTaskTime()` accesses `partitionGroup` which is created in the constructor. Since we cannot know what reorderings a compiler does, it would be better to specify `initializeTaskTime()` as static and pass in `partitionGroup`. To make `initializeCommittedOffsets()` future-proof, it would also be good to also transform it to a static method.

cadonna reviewed (commented):
@mjsax Thanks for the PR

Here my comments

cpettitt-confluent reviewed (approved):
LGTM!

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -62,7 +57,7 @@
      */
     AbstractTask(final TaskId id,
-                 final Collection<TopicPartition> partitions,
+                 final Set<TopicPartition> partitions,
```
Why reducing the type generality here?

mjsax replied:
In `StreamTask` constructor we call `consumer.committed(...)` that takes a `Set<TopicPartitions>` -- hence, it's easier to pass in a `Set` from the beginning on. Second, thinking about it, a `Collection` is actually too generic anyway -- we would not want to assign a `List` of partitions that may contain duplicates IMHO.

guozhangwang reviewed (commented):
Made a pass on the PR, lgtm modulo @cadonna 's comments.

mjsax committed:
Github comments

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java:
```diff
@@ -67,11 +67,13 @@ void initializeNewTasks() {
             try {
-                if (!entry.getValue().initializeStateStores()) {
+                final T task = entry.getValue();
+                task.initializeMetadata();
```
This is new -- we need to initialize the task after we created it.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -74,6 +79,9 @@
     }
 
+    @Override
+    public void initializeMetadata() {}
```
new method

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -325,6 +372,7 @@ public void resume() {
                 throw new ProcessorStateException(format("%sError while deleting the checkpoint file", logPrefix), e);
             }
         }
+        initializeMetadata();
```
This is also new -- we implicitly re-initialize the task on `resume()` -- was done "outside" before

guozhangwang replied:
Do we really need to re-initialize committed offsets and task time during Resume?

mjsax replied:
Yes, because in `StreamTask#suspend()` we call `closeTopology()` that calls `partitionGroup.clear()` that resets all times to `UNKNOWN`.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -164,9 +164,9 @@ public void punctuate(final long timestamp) {
-    static ProcessorTopology withRepartitionTopics(final List<ProcessorNode> processorNodes,
-                                                   final Map<String, SourceNode> sourcesByTopic,
-                                                   final Set<String> repartitionTopics) {
+    private static ProcessorTopology withRepartitionTopics(final List<ProcessorNode> processorNodes,
```
side cleanup

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -233,7 +233,7 @@ public void shouldHandleInitTransactionsTimeoutExceptionOnCreation() {
         final ProcessorTopology topology = withSources(
             asList(source1, source2, processorStreamTime, processorSystemTime),
-            mkMap(mkEntry(topic1, (SourceNode) source1), mkEntry(topic2, (SourceNode) source2))
+            mkMap(mkEntry(topic1, source1), mkEntry(topic2, source2))
```
side cleanup

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -353,7 +353,6 @@ private void assertTimeoutErrorLog(final LogCaptureAppender appender) {
         assertThat(expectedError, is(singletonList("ERROR")));
     }
 
-    @SuppressWarnings("unchecked")
```
side cleanup

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -890,11 +875,8 @@ public void shouldWrapKafkaExceptionsWithStreamsExceptionAndAddContextWhenPunctu
-                    throw new KafkaException("KABOOM!");
-                }
+            task.punctuate(processorStreamTime, 1, PunctuationType.STREAM_TIME, timestamp -> {
+                throw new KafkaException("KABOOM!");
```
side cleanup

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1333,6 +1302,41 @@ public void shouldWrapProducerFencedExceptionWithTaskMigragedExceptionInSuspendW
     }
 
+    @Test
+    public void shouldInitTaskTimeOnResumeWithEosDisabled() {
```
new test

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1577,20 +1576,23 @@ public void punctuate(final long timestamp) {
     public void shouldThrowProcessorStateExceptionOnInitializeOffsetsWhenAuthorizationException() {
         final Consumer<byte[], byte[]> consumer = mockConsumerWithCommittedException(new AuthorizationException("message"));
         final StreamTask task = createOptimizedStatefulTask(createConfig(false), consumer);
+        task.initializeMetadata();
```
we moved initializing the offsets from `initializeStateStores` into `initializeMetadata()`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1603,11 +1605,27 @@ public void shouldThrowWakeupExceptionOnInitializeOffsetsWhenWakeupException() {
 
         final ProcessorTopology topology = ProcessorTopologyFactories.with(
-            asList(source1),
+            singletonList(source1),
```
side cleanup

mjsax commented:
Updated this PR. Call for review.

As an afterthought, I realized that we should not init the offsets and task time within the constructor of `StreamTask` because we should not have a blocking call (`consumer.committed()`) during rebalance phase (note, that tasks are created in `onPartitionAssigned()` callback).

Hence, I refactored the code and added a new unified init method to get the offsets. This also addressed @cadonna concerns about the initialization order within the constructor.

guozhangwang commented:
@cadonna @cpettitt-confluent could you guys take another look?

cpettitt-confluent reviewed (approved):
One comment above on one of the side cleanups where I think we can go a bit further. Otherwise LGTM.

mjsax commented:
Java 11 / 2.12 passed. Others failed -- test result not available any longer.

Retest this please.

mjsax commented:
Java 11 / 2.13: `kafka.log.LogCleanerIntegrationTest.testIsThreadFailed`
Java 11 / 2.12: `org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest.shouldWorkWithRebalance`
Java 8 passed.

mjsax commented:
Merged to `trunk` and cherry-picked to `2.4`.
